### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by pbauer**

### CODE REVIEW

File: src/plone/rest/testing.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def setUpZope(self, app, configurationContext):
 class InternalServerErrorService(Service):
 
     def __call__(self):
-        from urllib2 import HTTPError
+        from six.moves.urllib.error import HTTPError
```
Comment (davisagli) [inline src/plone/rest/testing.py]: We should declare the dependency on `six` in setup.py

File: src/plone/rest/tests/test_permissions.py [hunk 2]
```diff
@@ -41,8 +41,9 @@ def traverse(self, path='/plone', accept='application/json', method='GET'):
         request.environ['PATH_TRANSLATED'] = path
         request.environ['HTTP_ACCEPT'] = accept
         request.environ['REQUEST_METHOD'] = method
-        request._auth = 'Basic %s' % b64encode(
-            '%s:%s' % (TEST_USER_NAME, TEST_USER_PASSWORD))
+        auth = '%s:%s' % (TEST_USER_NAME, TEST_USER_PASSWORD)
```
Comment (hvelarde) [inline src/plone/rest/tests/test_permissions.py]: shouldn't we use `.format()` instead?

File: src/plone/rest/tests/test_permissions.py [hunk 3]
```diff
@@ -41,8 +41,9 @@ def traverse(self, path='/plone', accept='application/json', method='GET'):
...
+        b64auth = b64encode(auth.encode('utf8'))
+        request._auth = 'Basic %s' % b64auth.decode('utf8')
```
Comment (hvelarde) [inline src/plone/rest/tests/test_permissions.py]: shouldn't we use `.format()` instead?

File: src/plone/rest/tests/test_traversal.py [hunk 4]
```diff
@@ -32,8 +32,9 @@ def traverse(self, path='/plone', accept='application/json', method='GET'):
         request.environ['PATH_TRANSLATED'] = path
         request.environ['HTTP_ACCEPT'] = accept
         request.environ['REQUEST_METHOD'] = method
-        request._auth = 'Basic %s' % b64encode(
-            '%s:%s' % (SITE_OWNER_NAME, SITE_OWNER_PASSWORD))
+        auth = '%s:%s' % (SITE_OWNER_NAME, SITE_OWNER_PASSWORD)
```
Comment (hvelarde) [inline src/plone/rest/tests/test_traversal.py]: shouldn't we use `.format()` instead?

File: src/plone/rest/tests/test_traversal.py [hunk 5]
```diff
@@ -32,8 +32,9 @@ def traverse(self, path='/plone', accept='application/json', method='GET'):
...
+        b64auth = b64encode(auth.encode('utf8'))
+        request._auth = 'Basic %s' % b64auth.decode('utf8')
```
Comment (hvelarde) [inline src/plone/rest/tests/test_traversal.py]: shouldn't we use `.format()` instead?

**Comment by hvelarde** [comment]
this seems to be almost ready now:

```console
$ pylint --py3k --disable=no-absolute-import src/plone/rest/
No config file found, using default configuration

------------------------------------
Your code has been rated at 10.00/10

```

**Comment by lukasgraf** [comment]
I've looked into the remaining 8 failures on Plone 5.2 / `py3.cfg`.

These aren't Python 3 failures as such, but come from the difference between `ZServer.ZPublisher` and `ZPublisher.WSGIPublisher`.

7 of the 8 failures are because status `401` doesn't get returned where it's expected:

```
AssertionError: The following assertions failed:

Request:  ('/', 'POST', ())
Expected: 401
Got:      404
...
```

 This is because in WSGIPublisher, [the `response._unauthorized()` gets triggered before the exception view is called](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/WSGIPublisher.py#L152-L162). In `ZPublisher` this happens the other way round, and `plone.rest` gets a change to [prevent PAS from issuing a challenge](https://github.com/plone/plone.rest/blob/7974081a7f247893e0a42fd0a3ac6fe3cfa4c983/src/plone/rest/errors.py#L38-L41) and triggering a redirect to the login form.

---

The remaining failure is caused by [`WSGIPublisher` normalizing HTTP exceptions](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/WSGIPublisher.py#L137-L141). This leads to the last actual exception (`sys.exc_info()`) being different from the exception that the error view gets, and causing `plone.rest` to fail on [this check](https://github.com/plone/plone.rest/blob/7974081a7f247893e0a42fd0a3ac6fe3cfa4c983/src/plone/rest/errors.py#L66-L69):

```
AssertionError: 'ERROR: Another exception happened before we could render the traceback.' is not an instance of <class 'list'>
```

**Comment by pbauer** [comment]
@davisagli or @tschorr could one of you have a look at this? I'm notoriously bad when it comes to understanding the various layers of exception-handling.

**Comment by tschorr** [comment]
>    The remaining failure is caused by [WSGIPublisher normalizing HTTP exceptions](https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/WSGIPublisher.py#L137-L141).

Always when I see `x-swgiorg.throw_errors` in code (as in the 2 lines above the Unathorized handling), I wonder wether this is used anywhere outside of Zope. The [specification](https://wsgi.readthedocs.io/en/latest/specifications/throw_errors.html) for this was created in 2006 and is still in proposed state ...

Maybe get rid of `x-wsgiorg.throw_errors` altogether?

**Comment by tschorr** [comment]
Running `bin/test -s plone.rest` with https://github.com/zopefoundation/Zope/pull/294 yields 3 failures and 1 error:

```
Failure in test test_401_unauthorized (plone.rest.tests.test_error_handling.TestErrorHandling)
Failure in test test_404_not_found (plone.rest.tests.test_error_handling.TestErrorHandling)
Failure in test test_500_internal_server_error (plone.rest.tests.test_error_handling.TestErrorHandling)
```

all fail with:

```
AssertionError: 'text/html;charset=utf-8' != 'application/json'
- text/html;charset=utf-8
+ application/json
 : When sending a GET request with Accept: application/json the server should respond with sending back application/json.
```

and then there's

```
Error in test test_500_traceback_only_for_manager_users (plone.rest.tests.test_error_handling.TestErrorHandling)
Traceback (most recent call last):
  File "/home/thomas/.pyenv/versions/3.6.5/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home/thomas/.pyenv/versions/3.6.5/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/thomas/devel/plone/buildout.coredev/src/plone.rest/src/plone/rest/tests/test_error_handling.py", line 107, in test_500_traceback_only_for_manager_users
    self.assertNotIn(u'traceback', response.json())
  File "/home/thomas/.buildout/eggs/cp36m/requests-2.19.1-py3.6.egg/requests/models.py", line 896, in json
    return complexjson.loads(self.text, **kwargs)
  File "/home/thomas/.buildout/eggs/cp36m/simplejson-3.16.0-py3.6-linux-x86_64.egg/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/home/thomas/.buildout/eggs/cp36m/simplejson-3.16.0-py3.6-linux-x86_64.egg/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/home/thomas/.buildout/eggs/cp36m/simplejson-3.16.0-py3.6-linux-x86_64.egg/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by davisagli** [comment]
@tschorr that's because you disabled exception view rendering, which is how plone.rest turns exceptions into JSON responses.

I would need to take a closer look at how to fix the problems that @lukasgraf pointed out, but getting rid of exception view handling is not it.

**Comment by tschorr** [comment]
>   This is because in WSGIPublisher, the response._unauthorized() gets triggered before the exception view is called. In ZPublisher this happens the other way round, and plone.rest gets a change to prevent PAS from issuing a challenge and triggering a redirect to the login form.

@lukasgraf I have reversed the order according to your proposal in https://github.com/zopefoundation/Zope/tree/py3_fix_wsgi_unauthorized and confirm this fixes 7 tests. Tests in Zope all pass with this modification.

**Comment by tschorr** [comment]
Jenkins looks slightly better using `py3_fix_wsgi_unauthorized`: https://jenkins.plone.org/view/PLIPs/job/plip-py3/582

**Comment by tschorr** [comment]
Checks for this PR pass now because only Python 2.7 is tested with Travis. I still have the one failing test mentioned by @lukasgraf ("Another exception ...") using Python 3.6.

**Comment by tisto** [comment]
@tschorr feel free to add Python 3 to the Travis configuration.

**Comment by lukasgraf** [comment]
@tschorr thanks for looking into this!

Just to clarify, I'm not necessarily *proposing* that the order in `WSGIPublisher` needs to be changed - I just noticed that this is the underlying cause of these failures in `plone.rest`. I left some feedback in [my comment on #304](https://github.com/zopefoundation/Zope/pull/304#issuecomment-406579566).

File: src/plone/rest/errors.py [hunk 6]
```diff
@@ -8,6 +8,11 @@
 from six.moves.urllib.parse import quote
 from six.moves.urllib.parse import unquote
 from zExceptions import NotFound
+try:
+    from ZPublisher.HTTPRequest import WSGIRequest
+    HAS_WSGI = True
+except:
```
Comment (hvelarde) [inline src/plone/rest/errors.py]: this is wrong, please add an exception here.

**Comment by frapell** [comment]
@pbauer I see this is hitting https://github.com/plone/plone.protect/issues/74 even though it is using python 2.7.14 (see latest travis https://travis-ci.org/plone/plone.rest/jobs/436770003)

Do we need a new release of plone.protect with the latest changes for tests to pass?

**Comment by tisto** [comment]
@jensens FYI: you merged this PR even though the Travis build failed. Please keep in mind that plone.rest and plone.restapi are used outside of Plone and for older Plone versions. It is just not sufficient if coredev tests pass.

**Pull request merged**